Share tag interfaces for tags

Nyholm opened the pull request:
We should share interfaces for tags with Symfony and the PHP community. This PR will add tag two interfaces for tags and implement these. 

Ping @nicolas-grekas: Can you give thumbs up on the interfaces? This is the interfaces I will write in a PSR.

Nyholm commented on src/Adapter/Common/TagAwareItem.php:
```diff
@@ -0,0 +1,35 @@
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface TagAwareItem
```
There is no need to extend CacheItemInterface here because items are not transferable between pools

Nyholm commented on src/Adapter/Common/TagAwarePool.php:
```diff
@@ -0,0 +1,34 @@
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+interface TagAwarePool extends CacheItemPoolInterface
```
We need to extend CacheItemPoolInterface inorder to make it type safe. 

Should we override the doc block for `getItem` and `getItems`?

nicolas-grekas commented:
See https://github.com/symfony/symfony/issues/19728#issuecomment-251674712 and following comments.
Looks like getTags on items is needed.
Interface suffix missing on the new item interface.

Nyholm commented:
Thank you for the review. I've updated the interfaces. 

Im comfortable with these interfaces (or something backwards compatible) will make it into the Fig. 

I've also updated all our adapters to support tagging. This PR is ready for a final review.

Nyholm commented:
The individual tests are failing because BC breaks in the common package. One need to merge this and bump a new version of cache/common and then the test will pass.

nicolas-grekas commented:
Did read the implementation details, but if you're fine the interfaces, so am I :)
I invite you to borrow the tag test cases in https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Tests/Adapter/TagAwareAdapterTest.php if you did not already do it (and report any deviation you notice if any).
